[Adrian Tan] iP

adriantan101 opened the pull request:
# DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

atjy5421 committed:
renamed Duke to Ruby and added firts message

atjy5421 committed:
ruby now echoes commands and exits when using bye

atjy5421 committed:
added the list feature

atjy5421 committed:
created Task.java and added ability to mark/unmark tasks

atjy5421 committed:
added Todo.java, Deadline.java and Event.java to handle different types of Tasks

atjy5421 committed:
Added testing using I/O redirection technique

atjy5421 committed:
added RubyException.java and handling exceptions

atjy5421 committed:
added deleting tasks

atjy5421 committed:
Make style changes and implement save function

atjy5421 committed:
Merge branch 'branch-Level-7'

atjy5421 committed:
Add functionality to chatbot to understand date and time

atjy5421 committed:
Merge branch 'branch-Level-8'

atjy5421 committed:
Applied OOP Principles

atjy5421 committed:
Remove showGoodbye() method and comments

atjy5421 committed:
Merge branch 'branch-A-MoreOOP'

atjy5421 committed:
Organize files into packages

atjy5421 committed:
Merge branch 'branch-A-Packages'

atjy5421 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

atjy5421 committed:
Integrate gradle into project

atjy5421 committed:
Merge branch 'branch-A-Gradle'

atjy5421 committed:
Create StorageTest.java and TodoTest.java

atjy5421 committed:
Merge branch 'branch-A-JUnit'

atjy5421 committed:
Add JavaDoc comments to all files

atjy5421 committed:
Add functionality to search for tasks based on keyword

atjy5421 committed:
Merge branch 'branch-A-JavaDoc'

atjy5421 committed:
Merge branch 'branch-Level-9'

Dominic-Khoo commented on src/main/java/myapp/task/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+ * about the date and time by which the task must be completed.
+ */
+public class Deadline extends Task {
+    private LocalDateTime by;
```
Perhaps a more specific name could be used for this variable? A possible suggestion could be dueDate.

Dominic-Khoo commented on src/main/java/myapp/command/Parser.java:
```diff
@@ -0,0 +1,233 @@
+
+        return switch (keyword) {
+            case "bye" -> "Bye. Hope to see you again soon!";
+            case "list" -> taskList.listTasks();
+            case "mark" -> handleMark(taskList, storage, words);
+            case "unmark" -> handleUnmark(taskList, storage, words);
+            case "todo" -> handleTodo(taskList, storage, words);
+            case "deadline" -> handleDeadline(taskList, storage, words);
+            case "event" -> handleEvent(taskList, storage, words);
+            case "delete" -> handleDelete(taskList, storage, words);
+            case "find" -> handleFind(taskList, words);
```
I like the naming here, where all methods are verbs 👍

Dominic-Khoo commented on src/main/java/myapp/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+     * @param tasks The list of tasks to be saved to the file.
+     * @throws IOException If an I/O error occurs while writing to the file.
+     */
+    public void save(List<Task> tasks) throws IOException {
```
Perhaps the method name could elaborate more on what it is saving?

Dominic-Khoo commented on src/main/java/myapp/command/Parser.java:
```diff
@@ -0,0 +1,233 @@
+            throw new RubyException("OOPS!!! The description of an event must include start and end times.");
+        }
+
+        String[] times = eventParts[1].split(" /to ");
```
I like the usage of plural variable names to indicate multi-valued variables!

Dominic-Khoo reviewed (commented):
LGTM! Overall naming conventions for code quality are mostly adhered to 👍

frymash commented on src/main/java/myapp/ui/Ui.java:
```diff
@@ -0,0 +1,60 @@
+        System.out.println(error);
+    }
+
+    /**
```
Good job for the JavaDoc comments you've written 👍

frymash commented on src/main/java/myapp/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+import myapp.task.*;
```
You could try importing the specific libraries/functions from myapp.task to avoid the need for wildcard imports.

frymash replied:
```suggestion
```
Another wildcard import to be removed.

frymash commented on src/main/java/myapp/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+                task = new Event(parts[2], from, to);
+                break;
+            default:
+                throw new RubyException("Invalid task type.");
```
[There's an interesting extension to switch expressions which you could consider using to eliminate the need for the `break` keyword in each case.](https://openjdk.org/jeps/361#:~:text=We%20extend%20the%20switch%20statement,uses%20only%20a%20single%20println%20.&text=In%20the%20common%20case%2C%20a,%3B%20default%20%2D%3E%20e3%3B%20%7D%3B)

frymash commented on src/main/java/myapp/task/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+            }
+        }
+        if (taskNumber == 1) {
+            sb.append("     No matching tasks found.");
```
Consider defining an indent constant since the indent here is reused in lines 95 and 112 as well.

frymash commented on src/test/java/myapp/storage/StorageTest.java:
```diff
@@ -0,0 +1,93 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
```suggestion
```

Perhaps this wildcard import could be removed as well.

frymash commented on src/main/java/myapp/command/Parser.java:
```diff
@@ -0,0 +1,233 @@
+        String[] words = command.split(" ", 2);
+        String keyword = words[0];
+
+        return switch (keyword) {
```
I think it's great how all the operations for each case are abstracted into separate functions so that the switch statement doesn't look so cluttered.

frymash commented on src/main/java/myapp/command/Parser.java:
```diff
@@ -0,0 +1,233 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import myapp.storage.Storage;
+import myapp.task.*;
```
```suggestion
```
Could consider removing the wildcard import here since it is a coding standard violation.

frymash commented on src/main/java/myapp/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+        for (String line : lines) {
+            String[] parts = line.split(" \\| ");
+            Task task;
+            switch (parts[0]) {
```
The switch statement could be trimmed a little by using the extended switch expression syntax (which you used in  lines 38-46 of Parser.java).

frymash reviewed (commented):
Apart from the minor nits surrounding code reuse and wildcard imports, your code is readable (partially thanks to your comprehensive use of JavaDoc) and adheres to coding standards. LGTM :)

JoanneHing commented on src/main/java/myapp/command/Parser.java:
I like how you separate the command handling logic into separate methods! well done!

JoanneHing commented on src/main/java/myapp/task/Event.java:
Maybe can considering creating another method that need the private field of the class instead of setting getters to return the private fields ?

JoanneHing commented on src/main/java/myapp/task/Task.java:
Same review as for the event class, could be applicable to the deadline class too!

JoanneHing reviewed (commented):
Overall LGTM! Just that maybe some of the getters methods can be reduced by creating new methods that handle the usage of the private fields! The general code is well done! Good job!

atjy5421 committed:
Use JavaFX to implement GUI

atjy5421 committed:
Merge branch 'branch-Level-10'

atjy5421 committed:
Implement assertions in task handling and UI

Assertions are not used to enforce and document key assumptions within the task management and UI code. This can lead to unhandled errors or invalid states, especially when dealing with malformed input or unexpected conditions.

To enhance code reliability and maintainability, we introduce assertions to ensure that critical assumptions are met during runtime. These include correctness of input data and task formats in the storage and task handling classes, as well as the validity of task data, including deadlines and events.

The use of assertions helps catch bugs during development by enforcing conditions that should always hold true. They are not intended for user input validation but serve as a safeguard to detect programming errors early. This approach improves code robustness and makes it easier to diagnose issues when they arise.

atjy5421 committed:
Refactor Parser, Ruby, and Storage classes to improve code quality

The codebase contains long methods, deep nesting, magic numbers, and complicated expressions, reducing readability and maintainability. Multiple levels of abstraction exist within methods, making the execution paths less clear. Variables are sometimes reused for different purposes, and there is some code duplication, all of which affects the overall code quality.

The current state increases the risk of introducing bugs when making future modifications.

Refactor Parser, Ruby, and Storage classes by breaking down long methods into smaller, single-purpose functions, and reduce deep nesting. Replace magic numbers with constants and use helper methods to simplify complex expressions.

The refactoring makes the happy path clearer and more prominent by reducing nesting and improving the separation of concerns making it easier for future changes. The code is also easier to test and maintain.

No logic has been altered during this refactoring, ensuring functionality remains intact.

atjy5421 committed:
data modified in tasks.txt

adriantan101 committed:
Merge pull request #2 from adriantan101/branch-A-CodeQuality

Improvements to code quality

adriantan101 committed:
Merge branch 'master' into branch-A-Assertions

adriantan101 committed:
Merge pull request #3 from adriantan101/branch-A-Assertions

Implement assertions in task handling and UI

atjy5421 committed:
Merge branch 'master' of https://github.com/adriantan101/ip

* 'master' of https://github.com/adriantan101/ip:
  data modified in tasks.txt
  Refactor Parser, Ruby, and Storage classes to improve code quality
  Implement assertions in task handling and UI

atjy5421 committed:
Fix bugs in the Storage.java

atjy5421 committed:
Add functionality to sort tasks alphabetically

atjy5421 committed:
Merge branch 'branch-C-Sort'

atjy5421 committed:
Update styling of GUI

atjy5421 committed:
Merge branch 'branch-CSS'

atjy5421 committed:
Add user guide in README

atjy5421 committed:
Update README with more detailed descriptions

atjy5421 committed:
Add image to README

atjy5421 committed:
Fix the issue with bye command not closing GUI

atjy5421 committed:
Add more images to README and a section for Getting Started

atjy5421 committed:
Changes to Storage.java for handling of data folder/file creation for jar file

atjy5421 committed:
Merge branch 'branch-jarfix'

atjy5421 committed:
Fix issues where past dates can be added with a deadline and event

atjy5421 committed:
Merge branch 'branch-eventdeadlinefix'

atjy5421 committed:
Change product screenshot in README

atjy5421 committed:
Add change to Task.java to satisfy Git convention

atjy5421 committed:
Add second change to Task.java to satisfy Git convention

atjy5421 committed:
Add third change to Task.java to satisfy Git convention

atjy5421 committed:
Add fourth change to Task.java to satisfy Git convention

atjy5421 committed:
Add last change to Task.java to satisfy Git convention
